core/src/main/java/com/datastax/oss/driver/internal/core/cql/PagingIterableSpliterator.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.core.cql;
+
+import com.datastax.oss.driver.api.core.PagingIterable;
+import com.datastax.oss.driver.shaded.guava.common.base.Preconditions;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.Objects;
+import java.util.Spliterator;
+import java.util.Spliterators;
+import java.util.function.Consumer;
+import net.jcip.annotations.NotThreadSafe;
+
+/**
+ * A Spliterator for {@link PagingIterable} instances that splits the stream in chunks of equal
+ * size.
+ *
+ * @param <ElementT> The element type of the underlying stream.
+ */
+@NotThreadSafe
+public class PagingIterableSpliterator<ElementT> implements Spliterator<ElementT> {
+
+  /**
+   * Creates a new {@link Builder} for {@link PagingIterableSpliterator} instances.
+   *
+   * @param iterable The {@link PagingIterable} to create a spliterator for.
+   * @param <ElementT> The iterable's element type
+   * @return a new {@link Builder}.
+   */
+  @NonNull
+  public static <ElementT> Builder<ElementT> builder(@NonNull PagingIterable<ElementT> iterable) {
+    return new Builder<>(iterable);
+  }
+
+  /** The default chunk size for {@link PagingIterableSpliterator}. */
+  public static final int DEFAULT_CHUNK_SIZE = 128;
+
+  private final PagingIterable<ElementT> iterable;
+  private long estimatedSize;
+  private final int chunkSize;
+  private final int characteristics;
+
+  /**
+   * Creates a new {@link PagingIterableSpliterator} for the given iterable, with unknown size and
+   * default chunk size ({@value #DEFAULT_CHUNK_SIZE}).
+   *
+   * @param iterable The {@link PagingIterable} to create a spliterator for.
+   */
+  public PagingIterableSpliterator(@NonNull PagingIterable<ElementT> iterable) {
+    this(iterable, Long.MAX_VALUE, DEFAULT_CHUNK_SIZE);
+  }
+
+  private PagingIterableSpliterator(
+      @NonNull PagingIterable<ElementT> iterable, long estimatedSize, int chunkSize) {
+    this.iterable = Objects.requireNonNull(iterable, "iterable cannot be null");
+    this.estimatedSize = estimatedSize;
+    this.chunkSize = chunkSize;
+    if (estimatedSize < Long.MAX_VALUE) {
+      characteristics =
+          Spliterator.ORDERED
```

Comment:
do we need to make our iterator `ORDERED`?
If the order is not important maybe we could remove that flag - I suspect that it may have a potential influence on the performance (the unordered version may be faster than ordered)

Reply:
I think `ORDERED` simply means that we get the elements in a deterministic order.

> Characteristic value signifying that an encounter order is defined for elements.

Reply:
Yes, `ORDERED` is different from `SORTED`.